src/librustc/middle/traits/project.rs

```diff
@@ -849,38 +849,39 @@ fn confirm_impl_candidate<'cx,'tcx>(
     impl_vtable: VtableImplData<'tcx, PredicateObligation<'tcx>>)
     -> (Ty<'tcx>, Vec<PredicateObligation<'tcx>>)
 {
+    let tcx = selcx.tcx();
     // there don't seem to be nicer accessors to these:
-    let impl_items_map = selcx.tcx().impl_items.borrow();
-    let impl_or_trait_items_map = selcx.tcx().impl_or_trait_items.borrow();
+    let impl_items_map = tcx.impl_items.borrow();
+    let impl_or_trait_items_map = tcx.impl_or_trait_items.borrow();
 
     let impl_items = &impl_items_map[impl_vtable.impl_def_id];
-    let mut impl_ty = None;
-    for impl_item in impl_items {
-        let assoc_type = match impl_or_trait_items_map[impl_item.def_id()] {
-            ty::TypeTraitItem(ref assoc_type) => assoc_type.clone(),
-            ty::MethodTraitItem(..) => { continue; }
-        };
-
-        if assoc_type.name != obligation.predicate.item_name {
-            continue;
-        }
-
-        let impl_poly_ty = ty::lookup_item_type(selcx.tcx(), assoc_type.def_id);
-        impl_ty = Some(impl_poly_ty.ty.subst(selcx.tcx(), &impl_vtable.substs));
-        break;
-    }
-
-    match impl_ty {
-        Some(ty) => (ty, impl_vtable.nested.into_vec()),
-        None => {
-            // This means that the impl is missing a
-            // definition for the associated type. This error
-            // ought to be reported by the type checker method
-            // `check_impl_items_against_trait`, so here we
-            // just return ty_err.
-            (selcx.tcx().types.err, vec!())
-        }
-    }
+    impl_items
+        .iter()
+        .filter_map(
+            |item| {
+                match impl_or_trait_items_map[item.def_id()] {
+                    ty::TypeTraitItem(ref assoc_type)
+                    if assoc_type.name == obligation.predicate.item_name => {
+                        Some(assoc_type)
+                    }
+                    _ => None
+                }
+            })
+        .next()
+        .map_or_else(
+            || {
+                // This means that the impl is missing a definition for the
+                // associated type. This error ought to be reported by the
+                // type checker method `check_impl_items_against_trait`, so
+                // here we just return ty_err.
+                (tcx.types.err, vec![])
+            },
+            move |assoc_type| {
+                let impl_poly_ty = ty::lookup_item_type(tcx, assoc_type.def_id);
+                let substs = selcx.ambient_substs().unwrap_or(&impl_vtable.substs);
```

Comment:
This does not feel right. For one thing, there is no obvious reason (to me) that it would make sense to apply the "ambient substs" to the impl. After all, these are the substitutions for the type parameters declared on the current fn, which is not the same as the impl we are projecting out of here. 

In any case, I think we should not need to pass them in at all, but rather we can apply them much earlier.